extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -148,50 +150,34 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
   }
 
   /** @override */
-  emitLifecycleEvent(eventName, opt_associatedEventData) {
+  emitLifecycleEvent(eventName) {
+    // This function can be invoked by the superclass constructor before this
+    // class's constructor gets to run.  In that case, this.lifecycleReporter_
+    // may not be initialized.  So initialize it here.  However, we'd rather
+    // not be doing this check and initialization on every emitLifecycleEvent
+    // call, so we'll pull a classic JS move and redefine this function to
+    // the underlying sendPing once we've initialized the lifecycleReporter_.
+    // Finally, make sure that the first ping actually gets sent.
     this.lifecycleReporter_ = this.lifecycleReporter_ ||
-        this.initLifecycleReporter();
-    switch (eventName) {
-      case 'adRequestEnd':
-        const fetchResponse = opt_associatedEventData;
-        const qqid = fetchResponse.headers.get(
-            this.lifecycleReporter_.QQID_HEADER);
-        this.lifecycleReporter_.setQqid(qqid);
-        break;
-      case 'adSlotCleared':
-        this.lifecycleReporter_.sendPing(eventName);
-        this.lifecycleReporter_.reset();
-        this.element.setAttribute('data-amp-slot-index',
-            this.win.ampAdSlotIdCounter++);
-        this.lifecycleReporter_ = this.initLifecycleReporter();
-        return;
-      case 'adSlotBuilt':
-      case 'urlBuilt':
-      case 'adRequestStart':
-      case 'extractCreativeAndSignature':
-      case 'adResponseValidateStart':
-      case 'renderFriendlyStart':
-      case 'renderCrossDomainStart':
-      case 'renderFriendlyEnd':
-      case 'renderCrossDomainEnd':
-      case 'preAdThrottle':
-      case 'renderSafeFrameStart':
-        break;
-      default:
-    }
-    this.lifecycleReporter_.sendPing(eventName);
+            this.initLifecycleReporter();
+    this.emitLifecycleEvent =
```

Comment:
This feels quite ugly to me.  Seems like this is the standard listener problem where its possible for events to occur before the listener to handle them is ready.  This is traditionally handled by queuing the events and then handing them off to the listener upon its construction.  So why don't we create a queue of events in A4A upon its initialization.  When the lifecycle reporter starts, it can be handed the queue and the queue within A4A can be destroyed such that later events do not attempt  to add themselves to the queue.  Happy to discuss offline.